### PR TITLE
【Hachathon No.30】

### PR DESCRIPTION
**Pull request by yangguohao**
### PR types
New features
### PR changes
APIs
### Describe
解决了issues：https://github.com/PaddlePaddle/Paddle/issues/40303
添加了paddle.nn.TripletMarginWithDistanceLoss以及paddle.nn.functional.triplet_margin_with_distance_loss的函数
设计文档：https://github.com/PaddlePaddle/community/pull/34
中文文档PR链接:https://github.com/PaddlePaddle/docs/pull/4644

### CODE REVIEW

**Comment by dingjiaweiww** [comment]
你的 PR 提交成功，感谢你对于开源项目的贡献

**Comment by dingjiaweiww** [comment]
PR 格式检查通过，你的PR 将接受Paddle 专家以及开源社区的review，请及时关注PR 动态

**Comment by shiyutang** [comment]
根据[流程](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/api_contributing_guides/api_contributing_guides_cn.html)，方案设计通过之后才能开始代码review，请仔细查看对应方案设计文档PR的修改意见，修改之后对应进行代码部分修改，当文档合入之后方可进行代码 PR 的review。

**Comment by shiyutang** [comment]
根据 CI 报错信息来看，你需要将paddle dev merge到当前 yangguohao:triplet_margin_distance_loss 分支来通过CI
image

**Review by shiyutang — CHANGES_REQUESTED** [review]
需要重新拉取develop的代码，通过所有CI（除PR-CI-Static-Check）之后才能进行进一步review

**Comment by dingjiaweiww** [comment]
请先通过CI噢~

**Comment by dingjiaweiww** [comment]
请先通过CI噢~

**Comment by shiyutang** [comment]
1. 需要增加设计文档链接，如图所示：
06b8a2904adf5ec3c229031619c25f57

2. 需要替换 in_dygraph_mode 如下图所示：
4ac43d3d67cdea50a08659a006f90a19

**Comment by shiyutang** [comment]
根据最新说明，_non_static_mode的调用路径需要更改为paddle.framework._non_static_mode，不再和fluid耦合

**Comment by shiyutang** [comment]
CI 中显示代码中存在错误：
image

**Comment by shiyutang** [comment]
请注意代码PR中还需要和中文文档关联才可合入，可以参考这个PR：https://github.com/PaddlePaddle/Paddle/pull/40472
image

书写规范请参考：https://github.com/PaddlePaddle/docs/wiki/%E9%A3%9E%E6%A1%A8API%E6%96%87%E6%A1%A3%E4%B9%A6%E5%86%99%E8%A7%84%E8%8C%83

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py [hunk 1]
```diff
@@ -0,0 +1,299 @@
+
```
Comment (shiyutang) [inline python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 空行删除

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (shiyutang) [inline python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 2022

File: python/paddle/nn/layer/loss.py [hunk 3]
```diff
@@ -1302,3 +1302,87 @@ def forward(self, input, label):
...
+            margins penalize cases where the negative examples are not distant enough from the
+            anchors, relative to the positives.
+    :param reduction:Indicate how to average the loss by batch_size.
+            the candicates are ``'none'`` | ``'mean'`` | ``'sum'``.
+            If :attr:`reduction` is ``'none'``, the unreduced loss is returned;
+            If :attr:`reduction` is ``'mean'``, the reduced mean loss is returned;
+            If :attr:`reduction` is ``'sum'``, the summed loss is returned.
+            Default: ``'mean'``
+    :return:Tensor. The tensor variable storing the triplet_margin_with_distance_loss of input and positive and negative.
+
```
Comment (shiyutang) [inline python/paddle/nn/layer/loss.py]: 注释需要根据自己理解重写，里面维度说明、positive、negative的数值说明也存在一定问题

File: python/paddle/nn/functional/loss.py [hunk 4]
```diff
@@ -2225,3 +2225,118 @@ def hinge_embedding_loss(input, label, margin=1.0, reduction='mean', name=None):
...
+
+    :param input:Input tensor, the data type is float32 or float64.
+            the shape is [N, \*], N is batch size and `\*` means any number of additional dimensions, available dtype is float32, float64.
+    :param positive:Positive tensor containing 1 or -1, the data type is float32 or float64.
+            The shape of label is the same as the shape of input.
+    :param negative:Negative tensor containing 1 or -1, the data type is float32 or float64.
+            The shape of label is the same as the shape of input.
+    :param distance_function: Quantifies the distance between two tensors. if not specified, 2 norm functions will be used.
+    :param swap:The distance swap is described in detail in the paperT
+            `Learning shallow convolutional feature descriptors with triplet losses` by
```
Comment (shiyutang) [inline python/paddle/nn/functional/loss.py]: 注释需要根据自己理解重写，里面维度说明、positive、negative的数值说明也存在一定问题

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+            reduction="unsupport reduction")
+        paddle.enable_static()
+
+    def test_TripletMarginDistanceLoss_distance_function(self):
+
+        def distance_function_1(x1, x2):
+            return 1.0 - paddle.nn.functional.cosine_similarity(x1, x2)
+
+        def distance_function_2(x1, x2):
+            return paddle.max(paddle.abs(x1-x2), axis=1)
```
Comment (shiyutang) [inline python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 测试当距离函数不满足非负性是是否会报错？

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+        expected = expected
+
+    return expected
+
+
+class TestTripletMarginWithDistanceLoss(unittest.TestCase):
+    def test_TripletMarginDistanceLoss(self):
+        input = np.random.uniform(0.1, 0.8, size=(20, 30)).astype(np.float64)
+        positive = np.random.randint(0, 2, size=(20, 30)).astype(np.float64)
+        negative = np.random.randint(0, 2, size=(20, 30)).astype(np.float64)
```
Comment (shiyutang) [inline python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 随机浮点数

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+                    place=place,
+                    input=input,
+                    positive=positive,
+                    negative=negative,
+                    reduction=reduction,
+                    functional=True)
+                self.assertTrue(np.allclose(static_functional, expected))
+                self.assertTrue(np.allclose(static_functional, dy_functional))
+                self.assertTrue(np.allclose(dy_functional, expected))
+
```
Comment (shiyutang) [inline python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py:232]: 增加margin<0, swap=True的测试。

**Review by shiyutang — CHANGES_REQUESTED** [review]
大部分完善点和 https://github.com/PaddlePaddle/Paddle/pull/40487#pullrequestreview-951140425 相同，请参照修改。

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py (hunk 5 above)
Comment (yangguohao) [reply python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 添加了一个对求得距离的判断不能小于0

**Review by shiyutang — APPROVED** [review]

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+        input = paddle.to_tensor([[0.1, 0.3], [1, 2]], dtype='float32')
+        positive = paddle.to_tensor([[0.0, 1.0]], dtype='float32')
+        negative = paddle.to_tensor([[0.2, 0.1]], dtype='float32')
+        self.assertRaises(
+            ValueError,
+            paddle.nn.functional.triplet_margin_with_distance_loss,
+            input=input,
+            positive=positive,
+            negative=negative, )
+        triplet_margin_with_distance_loss = paddle.nn.loss.TripletMarginWithDistanceLoss()
```
Comment (shiyutang) [inline python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 调用接口存在问题：
image

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+
+
+def call_TripletMarginDistanceLoss_layer(input,
+                    positive,
+                    negative,
+                    distance_function=None,
+                    margin=0.3,
+                    swap=False,
+                    reduction='mean',):
+    triplet_margin_with_distance_loss = paddle.nn.loss.TripletMarginWithDistanceLoss(distance_function=distance_function,
```
Comment (shiyutang) [inline python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 接口存在问题：
image

**Review by shiyutang — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py [hunk 10]
```diff
@@ -0,0 +1,362 @@
...
+                margin=0.3,
+                swap=False,
+                reduction='mean',
+                functional=False):
+    prog = paddle.static.Program()
+    startup_prog = paddle.static.Program()
+    with paddle.static.program_guard(prog, startup_prog):
+        input = paddle.fluid.data(
+            name='input', shape=input_np.shape, dtype='float64')
+        positive = paddle.fluid.data(
```
Comment (DDDivano) [inline python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 尽量避免使用fluid包下函数

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+
+        def distance_function_2(x1, x2):
+            return paddle.max(paddle.abs(x1-x2), axis=1)
+
+        distance_function_list = [distance_function_1,distance_function_2]
+        input = np.random.uniform(0.1, 0.8, size=(20, 30)).astype(np.float64)
+        positive = np.random.uniform(0, 2, size=(20, 30)).astype(np.float64)
+        negative = np.random.uniform(0, 2, size=(20, 30)).astype(np.float64)
+
+        place = fluid.CPUPlace()
```
Comment (DDDivano) [inline python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 使用paddle.CPUPlace()替代

**Review by DDDivano — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py (hunk 8 above)
Comment (yangguohao) [reply python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 我不太清楚这是什么原因导致的。

Comment (shiyutang) [reply python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 最好使用paddle.nn.TripletMarginWithDistanceLoss

**Comment by shiyutang** [comment]
coverage test fail 
image

File: python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py (hunk 9 above)
Comment (shiyutang) [reply python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 这里没有修改

Comment (yangguohao) [reply python/paddle/fluid/tests/unittests/test_triplet_margin_with_distance_loss.py]: 已修改

**Comment by yangguohao** [comment]
已修改相关内容，麻烦再review下

File: python/paddle/nn/functional/loss.py [hunk 12]
```diff
@@ -2225,3 +2225,123 @@ def hinge_embedding_loss(input, label, margin=1.0, reduction='mean', name=None):
...
+        L(input, pos, neg) = \max \{d(input_i, pos_i) - d(input_i, neg_i) + {\rm margin}, 0\}
+
+
+    where the default distance function
+
+    .. math::
+        d(x_i, y_i) = \left\lVert {\bf x}_i - {\bf y}_i \right\rVert_p
+
+    or user can defined their own distance functions.
+
```
Comment (Ligoml) [inline python/paddle/nn/functional/loss.py:2907]: 中英文文档需要保持一致，缺少论文部分

File: python/paddle/nn/functional/loss.py [hunk 13]
```diff
@@ -2225,3 +2225,123 @@ def hinge_embedding_loss(input, label, margin=1.0, reduction='mean', name=None):
...
+    Parameters:
+
+        input:Input tensor, the data type is float32 or float64.
```
Comment (Ligoml) [inline python/paddle/nn/functional/loss.py]: - 参数需要在括号里增加数据类型和是否可选
- 源码参数若有name则文档中也需要有，顺序与数量需要和源码保持一致

File: python/paddle/nn/layer/loss.py [hunk 14]
```diff
@@ -1302,3 +1302,91 @@ def forward(self, input, label):
             reduction=self.reduction,
             margin=self.margin,
             name=self.name)
+
+
+class TripletMarginWithDistanceLoss(Layer):
+    r"""
+    Creates a criterion that measures the triplet loss given an input
```
Comment (Ligoml) [inline python/paddle/nn/layer/loss.py:1407]: 参考triplet_margin_with_distance_loss进行修改

File: python/paddle/nn/functional/loss.py (hunk 12 above)
Comment (yangguohao) [reply python/paddle/nn/functional/loss.py:2907]: 已修改

File: python/paddle/nn/functional/loss.py (hunk 13 above)
Comment (yangguohao) [reply python/paddle/nn/functional/loss.py]: 已修改

File: python/paddle/nn/layer/loss.py (hunk 14 above)
Comment (yangguohao) [reply python/paddle/nn/layer/loss.py:1407]: 已修改

Comment (Ligoml) [reply python/paddle/nn/layer/loss.py:1407]: 嗯？介绍部分还是不太一样呀？
image
image

File: python/paddle/nn/layer/loss.py [hunk 15]
```diff
@@ -1302,3 +1302,93 @@ def forward(self, input, label):
...
+        reduction (str, Optional):Indicate how to average the loss by batch_size.
+                the candicates are ``'none'`` | ``'mean'`` | ``'sum'``.
+                If :attr:`reduction` is ``'none'``, the unreduced loss is returned;
+                If :attr:`reduction` is ``'mean'``, the reduced mean loss is returned;
+                If :attr:`reduction` is ``'sum'``, the summed loss is returned.
+                Default: ``'mean'``
+        name (str, optional): Name for the operation (optional, default is None).
+            For more information, please refer to :ref:`api_guide_Name`.
+	    
+    Call Parameters:
```
Comment (Ligoml) [inline python/paddle/nn/layer/loss.py]: - 这是？对应中文文档的shape吗？怎么没有output？
- 英文标点符号后最好加一个空格

File: python/paddle/nn/functional/loss.py (hunk 12 above)
Comment (Ligoml) [reply python/paddle/nn/functional/loss.py:2907]: 改了吗？
image
image

**Review by shiyutang — COMMENTED** [review]
LGTM

**Review by DDDivano — DISMISSED** [review]

File: python/paddle/nn/layer/loss.py (hunk 15 above)
Comment (yangguohao) [reply python/paddle/nn/layer/loss.py]: 将 Call Parameters 改为 Shapes， 添加 output

File: python/paddle/nn/layer/loss.py (hunk 14 above)
Comment (yangguohao) [reply python/paddle/nn/layer/loss.py:1407]: 添加英文部分 distance_function, margin, swap 的内容

File: python/paddle/nn/functional/loss.py (hunk 12 above)
Comment (yangguohao) [reply python/paddle/nn/functional/loss.py:2907]: 添加英文部分 distance_function, margin, swap 的内容

File: python/paddle/nn/layer/loss.py (hunk 14 above)
Comment (Ligoml) [reply python/paddle/nn/layer/loss.py:1407]: image

这里好像还没有加上~

File: python/paddle/nn/functional/loss.py (hunk 12 above)
Comment (Ligoml) [reply python/paddle/nn/functional/loss.py:2907]: 同 TripletMarginWithDistanceLoss，API介绍部分中英文意思还是不太一样

**Review by Ligoml — DISMISSED** [review]
LGTM for docs

File: python/paddle/nn/layer/loss.py [hunk 16]
```diff
@@ -1302,3 +1302,104 @@ def forward(self, input, label):
...
+            print(loss)
+            # Tensor([0.        , 0.57496738, 0.        ])
+
+            triplet_margin_with_distance_loss = TripletMarginWithDistanceLoss(reduction='mean')
+            loss = triplet_margin_with_distance_loss(input, positive, negative,)
+            print(loss)
+            # Tensor([0.19165580])
+
+    """
+    def __init__(self, distance_function=None, margin=1.0, swap=False, reduction: str = 'mean', name=None):
```
Comment (jeff41404) [inline python/paddle/nn/layer/loss.py]: the order of parameters here must same with [RFC](https://github.com/PaddlePaddle/community/pull/34), here suggest to modify RFC `distance_function` before `margin`, and class name `TripletMarginDistanceLoss` in RFC should be consistance with `TripletMarginWithDistanceLoss` here.

File: python/paddle/nn/functional/loss.py [hunk 17]
```diff
@@ -2225,3 +2225,132 @@ def hinge_embedding_loss(input, label, margin=1.0, reduction='mean', name=None):
...
     elif reduction == 'none':
         return loss
+
+
+def triplet_margin_with_distance_loss(input,
+				      positive,
+				      negative,
+				      distance_function = None,
+                        	      swap=False, 
+				      margin=1.0,
```
Comment (jeff41404) [inline python/paddle/nn/functional/loss.py]: the order of parameters here must same with [RFC](https://github.com/PaddlePaddle/community/pull/34), here suggest to modify postion of `margin` before `swap`

Comment (yangguohao) [reply python/paddle/nn/functional/loss.py]: 已修改

File: python/paddle/nn/layer/loss.py (hunk 16 above)
Comment (yangguohao) [reply python/paddle/nn/layer/loss.py:1385]: image

我这里显示的类名称和参数位置是正确的

Comment (yangguohao) [reply python/paddle/nn/layer/loss.py]: 已重提PR修改RFC

**Review by shiyutang — DISMISSED** [review]
LGTM，确认和RFC一致。

**Review by Ligoml — DISMISSED** [review]

**Review by jeff41404 — DISMISSED** [review]
LGTM

**Review by XiaoguangHu01 — DISMISSED** [review]
LGTM

**Review by Ligoml — DISMISSED** [review]

**Review by Ligoml — DISMISSED** [review]

**Review by Ligoml — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM
